### ISSUE TITLE
Rendering: incorrect size of circuit box

### ISSUE THREAD

**Issue by burlemarxiste** [Quandela/Perceval#414]
When a processor contains nested circuits (created with the `merge=False` argument passed to `add`), the blue box enclosing each subblock is of the wrong size.

The reason is that the drawing of elements stops at one level of recursion, whereas the `get_size` call goes to the full depth.

This is a circuit from a tutorial:

```python
import perceval as pcvl
from perceval.components import core_catalog
from perceval.rendering import format

from perceval.components.unitary_components import PS, BS, PERM

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, PS(phi=pcvl.P('phi_0')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_2')))
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, PS(phi=pcvl.P('phi_1')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_3')))
             .add(0, BS()))

chip = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=False)
              .add(0, upper_MZI, merge=False)
              .add(2, lower_MZI, merge=False))

processor = pcvl.Processor('SLOS', chip)
output_format = format.Format.MPLOT
extension = "pdf"

pcvl.pdisplay_to_file(processor, 'pdisplay_multiblocks.' + extension, output_format, recursive=True)
```

By changing some of the `merge=False` arguments to `True`, one can get circuits with various degrees of nesting. In all the following examples, the size of the blue circuit box is the same, and corresponds to the size of the fully flattened circuit.

<img width="1295" alt="Screenshot 2024-06-11 at 10 49 48" src="https://github.com/Quandela/Perceval/assets/171129116/b9b563b9-28ef-4801-846b-55b806715b32">
<img width="1263" alt="Screenshot 2024-06-11 at 10 50 27" src="https://github.com/Quandela/Perceval/assets/171129116/e36a8dd0-8a1e-46cc-94e8-feacf7e6fc81">
<img width="637" alt="Screenshot 2024-06-11 at 10 50 47" src="https://github.com/Quandela/Perceval/assets/171129116/d7aa3e17-4a1d-4251-a038-522a42d2c203">

In `ICircuitRenderer.render_circuit`, the call to `self.get_circuit_size(c, recursive=True)` is the culprit since it will always return the same value, `(9, 4)`, corresponding to the size of a completely flattened circuit.

Possible fixes:

1. Add a `recursion_level` parameter to all functions recursively traversing circuits (`flatten`, `get_size`, `get_circuit_size`, `render_circuit`, `pdisplay_processor`, `pdisplay_circuit`...), and make sure they consistently traverse the circuit graph at the appropriate depth.
2. Assume that one should never draw anything deeper than one level of recursion, and thus set `recursive=False` in the function call retrieving the size of a block. A similar change has to be done to the herald positioning code (it should `flatten` only one level deep to attach the heralds to the elements that are actually going to be drawn).
3. Assume that setting `recursive=True` should always go to the full recursion depth, at the risk of getting large diagrams. Thus `render_circuit` should pass its `recursive` argument when calling itself.

**Comment by oradomskyi** [comment]
Setting resursive=False in the render_circuit seems to be doing the trick. 

![image](https://github.com/user-attachments/assets/dc3d4efd-093c-4762-9038-0a462f303288)

![image](https://github.com/user-attachments/assets/334bf981-bb47-457c-beb5-d79770d595fa)

![image](https://github.com/user-attachments/assets/3a31447b-9186-4925-9bdf-5a023c530791)

![image](https://github.com/user-attachments/assets/d0710a32-06a2-4b91-b254-68a36a14e85a)

![image](https://github.com/user-attachments/assets/47036398-46f8-4005-8bd5-d47621bb3479)

![image](https://github.com/user-attachments/assets/cd1c8576-bd83-482b-9490-f6b7cb46c508)

![image](https://github.com/user-attachments/assets/bdbde225-538f-4425-ac24-3a39fd969c55)

![image](https://github.com/user-attachments/assets/cd5f3ff7-716c-474a-8ba3-17e4b9014366)

**Comment by oradomskyi** [comment]
corresponding txt and html files are correct as well

![image](https://github.com/user-attachments/assets/b2f5ef73-8a3a-4aa3-a7b9-54d6beadd4f6)

**Comment by oradomskyi** [comment]
![image](https://github.com/user-attachments/assets/d4ffe2be-f272-42cc-a59a-3ca3f22d7b2f)

![image](https://github.com/user-attachments/assets/b9cddcf0-e39a-4d91-b7b7-babc683a75bd)

**Comment by oradomskyi** [comment]
I will test circuits with more than 2 layers in the next days

**Comment by burlemarxiste** [comment]
Please do so! I believe there will be a problem with a circuit in which heralds are positioned around elements at a depth of 2: the function traversing the circuit to position them will process the full circuit (since it operates on the output of ```flatten()```, which doesn't stop at the first level), while the circuit will only be drawn down to a depth of 1, and thus the elements around which the heralds have been positioned won't be rendered, and the heralds will show up in unexpected positions, with missing lines.

You would need to add an argument to ```flatten``` to limit its recursion depth, and make sure that in ```collect_herald_info``` the circuit is only flattened down to a depth of 1.

This is the option **2** among the possible solutions I had highlighted when creating the issue. Maybe the maintainers of the library could chime in and confirm that it indeed is the preferred solution. From a previous exchange, it seems that only going down to one nesting level is a sane choice, since drawing at deeper levels would lead to big diagrams!

**Comment by oradomskyi** [comment]
Hi @burlemarxiste , could you please suggest a circuit for testing this? 

Well, Heralds are floating incorrectly anyway, so the problem is deeper than it looks. You can set different combinations of merge flags and see how it performs.

Setting the get_circuit_size(c, recursive=False) seem to make the circuit look better anyway.

```
import perceval as pcvl
from perceval.rendering import format
from perceval.components.unitary_components import PS, BS, PERM
from perceval import catalog

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, PS(phi=pcvl.P('phi_0')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_2')))
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, PS(phi=pcvl.P('phi_1')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_3')))
             .add(0, BS()))

lvl_0 = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=False)
              .add(0, upper_MZI, merge=True)
              .add(2, lower_MZI, merge=False))
              
pcvl.pdisplay(lvl_0)

lvl_1 = pcvl.Circuit(10,name="level_1").add(1, lvl_0, merge=True).add(5, lvl_0, merge=True)
lvl_2 = pcvl.Circuit(10, name="level_2").add(0, lvl_1, merge=True).add(0, lvl_1, merge=True)
lvl_3 = pcvl.Circuit(20, name="level_3").add(0, lvl_2, merge=True).add(10, lvl_2, merge=True)
pcvl.pdisplay(lvl_3)

p_lvl_3 = pcvl.Processor('SLOS', lvl_2)
p_lvl_3.add_herald(0,1)
p_lvl_3.add_herald(1,1)
p_lvl_3.add_herald(5,1)
p_lvl_3.add_herald(8,1)
p_lvl_3.add_herald(9,1)

output_format = format.Format.MPLOT
extension = "pdf"

pcvl.pdisplay_to_file(p_lvl_3, 'pdisplay_multiblocks.' + extension, output_format, recursive=True)
```

get_circuit_size(c, recursive=True)
![image](https://github.com/user-attachments/assets/a513a6c3-20c7-4d59-833d-c5a46aaa5ead)

get_circuit_size(c, recursive=False)
![image](https://github.com/user-attachments/assets/672d3cc5-2e3a-44f1-a8e6-60337d0ed278)

get_circuit_size(c, recursive=True)
![image](https://github.com/user-attachments/assets/59ef4a36-37bd-499a-800c-7c30bac27707)

get_circuit_size(c, recursive=False)
![image](https://github.com/user-attachments/assets/df0c87af-d348-4485-8452-8cb494ff5511)

**Comment by burlemarxiste** [comment]
The problem is caused because the circuit visited in `ˋ`collect_herald_infò`¨ is not the same as the circuit drawn. I think fixing ˋ`ˋflatten`ˋ` will help.

**Comment by oradomskyi** [comment]
Something is incorrect with indexing ports for heralds, horizontal and vertical positions of heralds.

```
import perceval as pcvl
from perceval.rendering import format
from perceval.components.unitary_components import PS, BS, PERM
from perceval import catalog

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, PS(phi=pcvl.P('phi_0')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_2')))
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, PS(phi=pcvl.P('phi_1')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_3')))
             .add(0, BS()))

lvl_0 = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=True)
              .add(0, upper_MZI, merge=False)
              .add(2, lower_MZI, merge=True))

lvl_1 = (pcvl.Circuit(10,name="Level1")
            .add(1, lvl_0, merge=True)
            .add(5, lvl_0, merge=True))

lvl_2 = (pcvl.Circuit(10, name="Level2")
            .add(0, lvl_1, merge=True).add(0, lvl_1, merge=True))

lvl_3 = (pcvl.Circuit(20, name="Level3")
            .add(0, lvl_2, merge=True).add(0, lvl_2, merge=True)
            .add(10, lvl_2, merge=True).add(10, lvl_2, merge=True))

p_lvl_3 = pcvl.Processor('SLOS', lvl_3)

for i in range(0,7):
    p_lvl_3.add_herald(i,0)

pcvl.pdisplay_to_file(p_lvl_3, 'pdisplay_multiblocks.pdf', format.Format.MPLOT, recursive=True)
```

![image](https://github.com/user-attachments/assets/7cf8f0f7-d0a6-4355-aba2-4ece69854f14)

**Comment by oradomskyi** [comment]
Herald positioning is a separate issue.. so what do you think about fixing the circuit box size issue via setting recursive=False?

**Comment by burlemarxiste** [comment]
The herald positioning issue seems to be caused here by what I described (with ```flatten```) plus another bug (the herald position info dict uses the circuit object as a key, which doesn't work when the same circuit reoccurs several time in the circuit). Give me a day to look into this!

**Comment by oradomskyi** [comment]
Hi @ericbrts what do you think about making heralds a if they were Ports and let the user add and configure them?

1. Make a herald a component
2. Let the user attach these to the processor and configure accordingly
3. Disable automatic generation of heralds - it looks to me like these add unnecessary complexity into rendering and traversing the nested circuits, as well as hidden dependency inside renderer.

Current implementation of automatic placement of heralds require refactoring as it breaks rendering of lines in the compound circuit in addition to not properly placing these heralds.

```
import perceval as pcvl
from perceval.rendering import format
from perceval.components.unitary_components import PS, BS, PERM
from perceval.components import Port
from perceval import catalog
from perceval.utils import Encoding, PostSelect

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, PS(phi=pcvl.P('phi_0')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_2')))
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, PS(phi=pcvl.P('phi_1')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_3')))
             .add(0, BS()))

lvl_0 = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=True)
              .add(0, upper_MZI, merge=False)
              .add(2, lower_MZI, merge=True))

lvl_1 = (pcvl.Circuit(10,name="Level1")
            .add(1, lvl_0, merge=True)
            .add(5, lvl_0, merge=True))

lvl_2 = (pcvl.Circuit(10, name="Level2")
            .add(0, lvl_1, merge=True).add(0, lvl_1, merge=True))

#lvl_3 = (pcvl.Circuit(20, name="Level3")
#            .add(0, lvl_2, merge=True).add(0, lvl_2, merge=True)
#            .add(10, lvl_2, merge=True).add(10, lvl_2, merge=False))

p = pcvl.Processor('SLOS', lvl_2)

p.set_postselection(PostSelect("[0,1]==1 & [2,3]==1"))
#p.set_postselection(PostSelect("[8,9]==1 & [10,11]==1"))

(p.add_port(0, Port(Encoding.DUAL_RAIL, 'Port_0_1'))
    .add_port(2, Port(Encoding.DUAL_RAIL, 'Port_2_3'))
    #.add_port(8, Port(Encoding.DUAL_RAIL, 'Port_8_9'))
    #.add_port(10, Port(Encoding.DUAL_RAIL, 'Port_10_11'))
)

for i in range(4, 7):
    p.add_herald(i,0)
#for i in range(12, 17):
#    p.add_herald(i,0)
    
pcvl.pdisplay_to_file(p, 'pdisplay_multiblocks.pdf', format.Format.MPLOT, recursive=True)
```

No heralds
![image](https://github.com/user-attachments/assets/c90f9737-aa78-4617-b5a0-eceef3feae84)

With heralds
![image](https://github.com/user-attachments/assets/54e92119-e8d0-4fef-9144-dd4cafbab6bf)

**Comment by oradomskyi** [comment]
Hi @burlemarxiste, let's wait what Eric says. I think fixing heralds in the way they are implemented now could make renderer even more intertwined with other classes and their properties :)

**Comment by oradomskyi** [comment]
Oh wow! I felt that would be something requiring more time to fix, great job @burlemarxiste ! 🥇😃

**Comment by burlemarxiste** [comment]
Note that your circuits make another part of the library choke:

```python
import perceval as pcvl
from perceval.rendering import format
from perceval.components.unitary_components import PS, BS, PERM
from perceval.components import Port
from perceval import catalog
from perceval.utils import Encoding, PostSelect

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, PS(phi=pcvl.P('phi_0')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_2')))
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, PS(phi=pcvl.P('phi_1')))
             .add(0, BS())
             .add(0, PS(phi=pcvl.P('phi_3')))
             .add(0, BS()))

lvl_0 = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=True)
              .add(0, upper_MZI, merge=False)
              .add(2, lower_MZI, merge=True))

lvl_1 = (pcvl.Circuit(10,name="Level1")
            .add(1, lvl_0, merge=True)
            .add(5, lvl_0, merge=True))

c = lvl_1.copy()
```

```console
RuntimeError: two parameters with the same name in the circuit (phi_0)
```

So it's possible to create a circuit with duplicate parameters names manually with calls to `add` but then this is flagged as forbidden when the circuit is copied.

**Comment by burlemarxiste** [comment]
https://github.com/burlemarxiste/Perceval/commit/a697ea0b07cff53c292b64af797fb0b9ae2ec3cc

This is the fix with which I could generate the correct renders.

Four toppings:
1. Your change (non-recursive ```get_size```).
2. Add an argument to ```flatten``` to optionally limit its depth.
3. Replace multiply occurring subcircuits by unique copies. I thought this was just a matter of calling ```copy```, but this required a rather ugly workaround for the issue I mentioned above (```copy()``` cannot copy a circuit if it contained reused circuits). @oradomskyi, if you have a better idea for this, let me know!
4. An extra check that sets the mode drawing style to ```PHOTONIC``` when heralds are not connected to anything. I'm not sure if this is really needed.

<img width="1220" alt="Screenshot 2024-07-31 at 16 13 18" src="https://github.com/user-attachments/assets/039c60d4-6b14-4fee-9ae2-6f07ad5f9b83">

**Comment by oradomskyi** [comment]
@burlemarxiste thanks for the update! I will take a look if a copy() could be made better.

**Comment by oradomskyi** [comment]
Hi @burlemarxiste , after quite a bit of debugging and watching the nested structure traversal I found the contradiction in the Perceval.

It does not allow different Parameter objects with the same name in the circuit, while it allows reference to the same Parameter object in multiple components... and that is OK to give a user uniform access to identical components:

https://github.com/Quandela/Perceval/blob/c5c087348d5fd3b3304677994779effc6f79bb8a/tests/test_circuit.py#L178C1-L195C1

What is not OK is that when there is a referenced Parameter object, and a user is calling circuit.copy() it invokes parameter.copy() and a Parameter makes a copy of itself so circuit copy fails.

Perceval is reusing Parameter objects, which allows to change the same property of multiple identical components present in the circuit.

I added `_is_copyable` parameter to the Parameter class, here is an example of building a multi-level circuit, copying it, and still having 'centralized' access to the desired parameters, see #447.

Let me know what you think,
Sa

P.S. As of now, I do not see any elegant way to copy a circuit together with these Parameter objects defined in the user scope, and provide user with handles to these objects.

```
import perceval as pcvl
from perceval.rendering import format
from perceval.components.unitary_components import PS, BS, PERM

my_phi0 = pcvl.P(name='phi_0', is_copyable=False)

my_PS0 = PS(phi=my_phi0)
my_PS1 = PS(phi=pcvl.P('phi_1', is_copyable=False))
my_PS2 = PS(phi=pcvl.P('phi_2', is_copyable=False))
my_PS3 = PS(phi=pcvl.P('phi_3', is_copyable=False))

pre_MZI = (pcvl.Circuit(4, name="Bell State Prep")
           .add(0, BS())
           .add(2, BS())
           .add(1, PERM([1, 0])))

upper_MZI = (pcvl.Circuit(2, name="upper MZI")
             .add(0, my_PS0)
             .add(0, BS())
             .add(0, my_PS2)
             .add(0, BS()))

lower_MZI = (pcvl.Circuit(2, name="lower MZI")
             .add(0, my_PS1)
             .add(0, BS())
             .add(0, my_PS3)
             .add(0, BS()))

lvl_0 = (pcvl.Circuit(4)
              .add(0, pre_MZI, merge=True)
              .add(0, upper_MZI, merge=True)
              .add(2, lower_MZI, merge=False))

lvl_1 = (pcvl.Circuit(10,name="Level1")
            .add(1, lvl_0, merge=True)
            .add(5, lvl_0, merge=True))

lvl_2 = (pcvl.Circuit(10, name="Level2")
            .add(0, lvl_1, merge=True).add(0, lvl_1, merge=True))

lvl_2_copy = lvl_2.copy()

p_lvl_2 = pcvl.Processor('SLOS', lvl_2)
p_lvl_2_copy = pcvl.Processor('SLOS', lvl_2_copy)

for angle in [15, 75]:
    my_phi0.set_value(angle)

    print('\r\nOriginal Processor', hex(id(p_lvl_2)), f"angle={angle}",my_phi0)
    pcvl.pdisplay(p_lvl_2, format.Format.MPLOT, recursive=True)
    print('\r\nCopied Processor', hex(id(p_lvl_2_copy)), f"angle={angle}",my_phi0)
    pcvl.pdisplay(p_lvl_2_copy, format.Format.MPLOT, recursive=True)

    print('\r\n\r\n')
```


![image](https://github.com/user-attachments/assets/5fd1e4e2-a29f-4340-b87b-ee37295a77f0)
![image](https://github.com/user-attachments/assets/471d21ff-6803-43b7-8a0a-ee6fe948a0db)
![image](https://github.com/user-attachments/assets/82bb9961-9e12-48ea-8757-10184835f822)
![image](https://github.com/user-attachments/assets/ceeaa8a7-220a-4626-b6b1-b86c0acc1e3c)

**Comment by ericbrts** [comment]
Hello,

This issue is actually dealing with 2 different bugs:
* The one with the box recursive display which you've just fixed @oradomskyi 
* Another one with "floating herald ports" which @raksharuia fixed in an integration branch.

Now that we've merged both fixes to the same branch (release/0.11.1), I think it'd be better to test the result on the release branch, and open a new issue should a bug still exist.

Anyway, thanks a lot for your help! :)
Eric

**Comment by oradomskyi** [comment]
Hi @ericbrts,

Thanks for the feedback!

You're right and the discussion here have diverged into a very different topic... 

Do you want me to move messages related to the issue of "circuit.copy() fails" to the #447? :)

Thanks,
Sasha
